File: locus-api-core/src/main/java/locus/api/objects/extra/TrackStats.java
package locus.api.objects.extra;

import java.io.IOException;

import locus.api.objects.Storable;
import locus.api.utils.DataReaderBigEndian;
import locus.api.utils.DataWriterBigEndian;

/**
 * Created by menion on 23. 6. 2014.
 * Asamm Software, s. r. o.
 */
public class TrackStats extends Storable {

    // number of points (useful when points itself are not loaded)
    private int mNumOfPoints;
    // track start time (time of first point)
    private long mStartTime;
    // track stop time (time of last point)
    private long mStopTime;

    // total length of done route
    private float mTotalLength;
    // total track distance with speed
    private float mTotalLengthMove;
    // total time of route
    private long mTotalTime;
    // total track time with speed
    private long mTotalTimeMove;
    // maximal speed of this route
    private float mSpeedMax;

    // maximum altitude on track
    private float mAltitudeMax;
    // minimum altitude on track
    private float mAltitudeMin;

    // neutral grade (distance)
    private float mEleNeutralDistance;
    // neutral grade (elevation)
    private float mEleNeutralHeight;
    // positive grade (distance)
    private float mElePositiveDistance;
    // positive grade (elevation)
    private float mElePositiveHeight;
    // negative grade (distance)
    private float mEleNegativeDistance;
    // negative grade (elevation)
    private float mEleNegativeHeight;
    // total grade (distance)
    @Deprecated
    private float mEleTotalAbsDistance;
    // total grade (elevation)
    @Deprecated
    private float mEleTotalAbsHeight;

    // number of measured beats
    private double mHeartRateBeats;
    // time during which were beats measured
    private long mHeartRateTime;
    // maximum HRM value
    private int mHeartRateMax;

    // number of measured cadence revolutions
    private double mCadenceNumber;
    // time during which were cadence measured
    private long mCadenceTime;
    // maximum cadence value
    private int mCadenceMax;

    // burned energy (in joule)
    private int mEnergy;

    // number of steps
    private int mNumOfStrides;

    /**
     * Default empty constructor.
     */
    public TrackStats() {
        mNumOfPoints = 0;
        mStartTime = -1L;
        mStopTime = -1L;

        // other variables
        resetStatistics();
    }

    //*************************************************
    // GETTERS & SETTERS
    //*************************************************

    // NUMBER OF POINTS

    /**
     * Number of points in track.
     *
     * @return number of trackpoints
     */
    public int getNumOfPoints() {
        return mNumOfPoints;
    }

    /**
     * Set number of trackpoints in current track.
     *
     * @param numOfPoints number of trackpoints
     */
    public void setNumOfPoints(int numOfPoints) {
        this.mNumOfPoints = numOfPoints;
    }

    // START TIME

    /**
     * Get time when current track started.
     *
     * @return time of start (in ms)
     */
    public long getStartTime() {
        return mStartTime;
    }

    /**
     * Set new value to start time parameter.
     *
     * @param startTime time when track started (in ms)
     */
    public void setStartTime(long startTime) {
        this.mStartTime = startTime;
    }

    // STOP TIME

    /**
     * Get current defined stop time of track.
     *
     * @return stop time
     */
    public long getStopTime() {
        return mStopTime;
    }

    /**
     * Set new time when track record was stopped (time of last trackpoint).
     *
     * @param stopTime stop time
     */
    public void setStopTime(long stopTime) {
        this.mStopTime = stopTime;
    }

    // TOTAL LENGTH

    public float getTotalLength() {
        return mTotalLength;
    }

    public void setTotalLength(float totalLength) {
        this.mTotalLength = totalLength;
    }

    public void addTotalLength(float add) {
        this.mTotalLength += add;
    }

    // TOTAL LENGTH (MOVE)

    public float getTotalLengthMove() {
        return mTotalLengthMove;
    }

    public void setTotalLengthMove(float totalLengthMove) {
        this.mTotalLengthMove = totalLengthMove;
    }

    public void addTotalLengthMove(float add) {
        this.mTotalLengthMove += add;
    }

    // TOTAL TIME

    public long getTotalTime() {
        return mTotalTime;
    }

    public void addTotalTime(long add) {
        this.mTotalTime += Math.abs(add);
    }

    public void setTotalTime(long totalTime) {
        this.mTotalTime = Math.abs(totalTime);
    }

    // TOTAL TIME (MOVE)

    public long getTotalTimeMove() {
        return mTotalTimeMove;
    }

    public void setTotalTimeMove(long totalTimeMove) {
        this.mTotalTimeMove = Math.abs(totalTimeMove);
    }

    public void addTotalTimeMove(long add) {
        this.mTotalTimeMove += Math.abs(add);
    }

    // SPEED MAX

    /**
     * Get maximal speed value received from GPS.
     *
     * @return maximal speed value (in m/s)
     */
    public float getSpeedMax() {
        return mSpeedMax;
    }

    /**
     * Set maximal speed value.
     *
     * @param speedMax max. speed value (in m/s)
     */
    public void setSpeedMax(float speedMax) {
        this.mSpeedMax = speedMax;
    }

    // ALTITUDE MAX

    public float getAltitudeMax() {
        return mAltitudeMax;
    }

    public void setAltitudeMax(float altitudeMax) {
        this.mAltitudeMax = altitudeMax;
    }

    // ALTITUDE MIN

    public float getAltitudeMin() {
        return mAltitudeMin;
    }

    public void setAltitudeMin(float altitudeMin) {
        this.mAltitudeMin = altitudeMin;
    }

    // ELEVATION NEUTRAL - DISTANCE

    public float getEleNeutralDistance() {
        return mEleNeutralDistance;
    }

    public void setEleNeutralDistance(float eleNeutralDistance) {
        this.mEleNeutralDistance = eleNeutralDistance;
    }

    public void addEleNeutralDistance(float add) {
        this.mEleNeutralDistance += add;
    }

    // ELEVATION NEUTRAL - HEIGHT

    public float getEleNeutralHeight() {
        return mEleNeutralHeight;
    }

    public void setEleNeutralHeight(float eleNeutralHeight) {
        this.mEleNeutralHeight = eleNeutralHeight;
    }

    /**
     * Add elevation value for "neutral height"
     *
     * @param add value increment
     */
    public void addEleNeutralHeight(float add) {
        this.mEleNeutralHeight += add;
    }

    // ELEVATION POSITIVE - DISTANCE

    public float getElePositiveDistance() {
        return mElePositiveDistance;
    }

    public void setElePositiveDistance(float elePositiveDistance) {
        this.mElePositiveDistance = elePositiveDistance;
    }

    public void addElePositiveDistance(float add) {
        this.mElePositiveDistance += add;
    }

    // ELEVATION POSITIVE - HEIGHT

    public float getElePositiveHeight() {
        return mElePositiveHeight;
    }

    public void setElePositiveHeight(float elePositiveHeight) {
        this.mElePositiveHeight = elePositiveHeight;
    }

    /**
     * Add elevation value for "positive height"
     *
     * @param add value increment
     */
    public void addElePositiveHeight(float add) {
        this.mElePositiveHeight += add;
    }

    // ELEVATION NEGATIVE - DISTANCE

    public float getEleNegativeDistance() {
        return mEleNegativeDistance;
    }

    public void setEleNegativeDistance(float eleNegativeDistance) {
        this.mEleNegativeDistance = eleNegativeDistance;
    }

    public void addEleNegativeDistance(float add) {
        this.mEleNegativeDistance += add;
    }

    // ELEVATION NEGATIVE - HEIGHT

    public float getEleNegativeHeight() {
        return mEleNegativeHeight;
    }

    public void setEleNegativeHeight(float eleNegativeHeight) {
        this.mEleNegativeHeight = eleNegativeHeight;
    }

    /**
     * Add elevation value for "negative height"
     *
     * @param add value increment
     */
    public void addEleNegativeHeight(float add) {
        this.mEleNegativeHeight += add;
    }

    // ELEVATION TOTAL - DISTANCE

    @Deprecated
    public float getEleTotalAbsDistance() {
        return mEleTotalAbsDistance;
    }

    public void setEleTotalAbsDistance(float eleTotalAbsDistance) {
        this.mEleTotalAbsDistance = eleTotalAbsDistance;
    }

    public void addEleTotalAbsDistance(float add) {
        this.mEleTotalAbsDistance += add;
    }

    // ELEVATION TOTAL - HEIGHT

    @Deprecated
    public float getEleTotalAbsHeight() {
        return mEleTotalAbsHeight;
    }

    public void setEleTotalAbsHeight(float eleTotalAbsHeight) {
        this.mEleTotalAbsHeight = eleTotalAbsHeight;
    }

    public void addEleTotalAbsHeight(float add) {
        this.mEleTotalAbsHeight += add;
    }

    // HEART RATE VALUES

    /**
     * Get average value of heart rate.
     *
     * @return average value (in BPM)
     */
    public int getHrmAverage() {
        if (mHeartRateBeats > 0 && mHeartRateTime > 0L) {
            double minutes = mHeartRateTime / (60.0 * 1000.0);
            return (int) (mHeartRateBeats / minutes);
        } else {
            return 0;
        }
    }

    /**
     * Get maximal value of heart rate.
     *
     * @return maximal heart rate value (in BPM)
     */
    public int getHrmMax() {
        return mHeartRateMax;
    }

    /**
     * Add measured heart rate values.
     *
     * @param hrmMeasured   measured heart rate value (in bpm)
     * @param hrmAvgSegment average heart rate value in segment (in bpm)
     * @param measureTime   time of segment (in millis)
     */
    public void addHeartRateMeasure(int hrmMeasured, int hrmAvgSegment, long measureTime) {
        // store values
        double inMinutes = measureTime * 1.0 / (60 * 1000);
        double numOfBeats = hrmAvgSegment * inMinutes;
        mHeartRateBeats += numOfBeats;
        mHeartRateTime += measureTime;

        // compute max. heart rate
        mHeartRateMax = Math.max(mHeartRateMax, hrmMeasured);
    }

    // CADENCE VALUES

    /**
     * Get average value of cadence.
     *
     * @return average cadence value (
     */
    public int getCadenceAverage() {
        if (mCadenceNumber > 0 && mCadenceTime > 0L) {
            double minutes = mCadenceTime / (60.0 * 1000.0);
            return (int) (mCadenceNumber / minutes);
        } else {
            return 0;
        }
    }

    /**
     * Get maximal value of heart rate.
     *
     * @return maximal heart rate value (in BPM)
     */
    public int getCadenceMax() {
        return mCadenceMax;
    }

    /**
     * Add measured cadence values.
     *
     * @param revMeasured   measured cadence value (in rpm)
     * @param revAvgSegment average revolutions value in segment (in rpm)
     * @param measureTime   time of segment (in millis)
     */
    public void addCadenceMeasure(int revMeasured, int revAvgSegment, long measureTime) {
        // store values
        double inMinutes = measureTime * 1.0 / (60 * 1000);
        double numOfRevolutions = revAvgSegment * inMinutes;
        mCadenceNumber += numOfRevolutions;
        mCadenceTime += measureTime;

        // compute max. heart rate
        mCadenceMax = Math.max(mCadenceMax, revMeasured);
    }

    // ENERGY BURNED

    /**
     * Get burned energy.
     *
     * @return burned energy (in joule)
     */
    public int getEnergy() {
        return mEnergy;
    }

    /**
     * Add burned energy.
     *
     * @param energy burned energy (in joule)
     */
    public void addEnergy(int energy) {
        mEnergy += energy;
    }

    // STRIDES

    /**
     * Get total number of strides.
     *
     * @return number of strides
     */
    public int getNumOfStrides() {
        return mNumOfStrides;
    }

    /**
     * Set total number of strides.
     *
     * @param numOfStrides number of strides
     */
    public void setNumOfStrides(int numOfStrides) {
        mNumOfStrides = numOfStrides;
    }

    //*************************************************
    // OTHER TOOLS
    //*************************************************

    public double getTrackLength(boolean onlyWithMove) {
        if (onlyWithMove) {
            return mTotalLengthMove;
        } else {
            return mTotalLength;
        }
    }

    public long getTrackTime(boolean onlyWithMove) {
        if (onlyWithMove) {
            return mTotalTimeMove;
        } else {
            return mTotalTime;
        }
    }

    /**
     * Get average speed value for a track.
     *
     * @param onlyWithMove {@code true} to get AVG. speed only during movement
     * @return average speed in m/s
     */
    public float getSpeedAverage(boolean onlyWithMove) {
        float trackTime = getTrackTime(onlyWithMove) / 1000.0f;
        if (trackTime > 0) {
            return (float) (getTrackLength(onlyWithMove) / trackTime);
        } else {
            return 0.0f;
        }
    }

    /**
     * Check if track has some elevation values.
     *
     * @return {@code true} if track has elevation values
     */
    public boolean hasElevationValues() {
        return (mAltitudeMin != Float.POSITIVE_INFINITY && mAltitudeMin != 0.0)
                || (mAltitudeMax != Float.NEGATIVE_INFINITY && mAltitudeMax != 0.0);
    }

    /**
     * Reset all parameters of track statistics.
     */
    public void resetStatistics() {
        // basic statistics variables
        mTotalLength = 0.0f;
        mTotalLengthMove = 0.0f;
        mTotalTime = 0L;
        mTotalTimeMove = 0L;
        mSpeedMax = 0.0f;
        mHeartRateBeats = 0.0;
        mHeartRateTime = 0L;
        mHeartRateMax = 0;
        mCadenceNumber = 0.0;
        mCadenceTime = 0L;
        mCadenceMax = 0;
        mEnergy = 0;
        mNumOfStrides = 0;

        // reset also elevation values
        resetStatisticsAltitude();
    }

    /**
     * Reset only statistics related to elevation values.
     */
    public void resetStatisticsAltitude() {
        mAltitudeMax = Float.NEGATIVE_INFINITY;
        mAltitudeMin = Float.POSITIVE_INFINITY;

        // graph variables
        mEleNeutralDistance = 0.0f;
        mEleNeutralHeight = 0.0f;
        mElePositiveDistance = 0.0f;
        mElePositiveHeight = 0.0f;
        mEleNegativeDistance = 0.0f;
        mEleNegativeHeight = 0.0f;
        mEleTotalAbsDistance = 0.0f;
        mEleTotalAbsHeight = 0.0f;
    }

    /**
     * Function that allows to merge more statistics into one.
     *
     * @param stats second statistics that will be merged into this.
     */
    public void appendStatistics(TrackStats stats) {
        this.mNumOfPoints += stats.mNumOfPoints;
        this.mStartTime = Math.min(mStartTime, stats.mStartTime);
        this.mStopTime = Math.max(mStopTime, stats.mStopTime);
        this.mTotalLength += stats.mTotalLength;
        this.mTotalLengthMove += stats.mTotalLengthMove;
        this.mTotalTime += stats.mTotalTime;
        this.mTotalTimeMove += stats.mTotalTimeMove;
        this.mSpeedMax = Math.max(mSpeedMax, stats.mSpeedMax);
        this.mAltitudeMax = Math.max(mAltitudeMax, stats.mAltitudeMax);
        this.mAltitudeMin = Math.min(mAltitudeMin, stats.mAltitudeMin);

        this.mEleNeutralDistance += stats.mEleNeutralDistance;
        this.mEleNeutralHeight += stats.mEleNeutralHeight;
        this.mElePositiveDistance += stats.mElePositiveDistance;
        this.mElePositiveHeight += stats.mElePositiveHeight;
        this.mEleNegativeDistance += stats.mEleNegativeDistance;
        this.mEleNegativeHeight += stats.mEleNegativeHeight;
        this.mEleTotalAbsDistance += stats.mEleTotalAbsDistance;
        this.mEleTotalAbsHeight += stats.mEleTotalAbsHeight;

        this.mHeartRateBeats += stats.mHeartRateBeats;
        this.mHeartRateTime += stats.mHeartRateTime;
        this.mHeartRateMax = Math.max(this.mHeartRateMax, stats.mHeartRateMax);

        this.mCadenceNumber += stats.mCadenceNumber;
        this.mCadenceTime += stats.mCadenceTime;
        this.mCadenceMax = Math.max(this.mCadenceMax, stats.mCadenceMax);

        this.mEnergy += stats.mEnergy;
        this.mNumOfStrides += stats.mNumOfStrides;
    }

    //*************************************************
    // STORABLE PART
    //*************************************************

    @Override
    protected int getVersion() {
        return 3;
    }

    @Override
    protected void readObject(int version, DataReaderBigEndian dr) throws IOException {
        mNumOfPoints = dr.readInt();
        mStartTime = dr.readLong();
        mStopTime = dr.readLong();

        // basic variables
        mTotalLength = dr.readFloat();
        mTotalLengthMove = dr.readFloat();
        mTotalTime = dr.readLong();
        mTotalTimeMove = dr.readLong();
        mSpeedMax = dr.readFloat();

        // altitude values
        mAltitudeMax = dr.readFloat();
        mAltitudeMin = dr.readFloat();

        // elevation variables
        mEleNeutralDistance = dr.readFloat();
        mEleNeutralHeight = dr.readFloat();
        mElePositiveDistance = dr.readFloat();
        mElePositiveHeight = dr.readFloat();
        mEleNegativeDistance = dr.readFloat();
        mEleNegativeHeight = dr.readFloat();
        mEleTotalAbsDistance = dr.readFloat();
        mEleTotalAbsHeight = dr.readFloat();

        // V1

        if (version >= 1) {
            mHeartRateBeats = dr.readInt();
            mHeartRateTime = dr.readLong();
            mHeartRateMax = dr.readInt();
            mEnergy = dr.readInt();
        }

        // V2

        if (version >= 2) {
            mHeartRateBeats = dr.readDouble();
            mCadenceNumber = dr.readDouble();
            mCadenceTime = dr.readLong();
            mCadenceMax = dr.readInt();
        }

        // V3

        if (version >= 3) {
            mNumOfStrides = dr.readInt();
        }
    }

    @Override
    protected void writeObject(DataWriterBigEndian dw) throws IOException {
        dw.writeInt(mNumOfPoints);
        dw.writeLong(mStartTime);
        dw.writeLong(mStopTime);

        // basic variables
        dw.writeFloat(mTotalLength);
        dw.writeFloat(mTotalLengthMove);
        dw.writeLong(mTotalTime);
        dw.writeLong(mTotalTimeMove);
        dw.writeFloat(mSpeedMax);

        // altitude values
        dw.writeFloat(mAltitudeMax);
        dw.writeFloat(mAltitudeMin);

        // elevation variables
        dw.writeFloat(mEleNeutralDistance);
        dw.writeFloat(mEleNeutralHeight);
        dw.writeFloat(mElePositiveDistance);
        dw.writeFloat(mElePositiveHeight);
        dw.writeFloat(mEleNegativeDistance);
        dw.writeFloat(mEleNegativeHeight);
        dw.writeFloat(mEleTotalAbsDistance);
        dw.writeFloat(mEleTotalAbsHeight);

        // V1

        dw.writeInt((int) mHeartRateBeats);
        dw.writeLong(mHeartRateTime);
        dw.writeInt(mHeartRateMax);
        dw.writeInt(mEnergy);

        // V2

        dw.writeDouble(mHeartRateBeats);
        dw.writeDouble(mCadenceNumber);
        dw.writeLong(mCadenceTime);
        dw.writeInt(mCadenceMax);

        // V3

        dw.writeInt(mNumOfStrides);
    }
}

File: locus-api-core/src/main/java/locus/api/objects/extra/TrackStats.kt
/**
 * Created by menion on 23. 6. 2014.
 * Asamm Software, s. r. o.
 */
package locus.api.objects.extra

import locus.api.objects.Storable
import locus.api.utils.DataReaderBigEndian
import locus.api.utils.DataWriterBigEndian
import java.io.IOException
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * Precise statistics for the [locus.api.objects.geoData.Track] class.
 */
class TrackStats : Storable() {

    /**
     * Number of points (useful when points itself are not loaded).
     */
    var numOfPoints: Int = 0

    /**
     * Track start time (time of first point) (in ms).
     */
    var startTime: Long = -1L

    /**
     * Track stop time (time of last point) (in ms).
     */
    var stopTime: Long = -1L

    // TOTAL LENGTH

    /**
     * Total length of done route (in m).
     */
    var totalLength: Float = 0.0f

    fun addTotalLength(add: Float) {
        this.totalLength += add
    }

    /**
     * Total track distance with speed (in m).
     */
    var totalLengthMove: Float = 0.0f

    fun addTotalLengthMove(add: Float) {
        this.totalLengthMove += add
    }

    fun getLength(onlyWithMove: Boolean): Float {
        return if (onlyWithMove) {
            totalLengthMove
        } else {
            totalLength
        }
    }

    // TOTAL TIME

    /**
     * Total time of route (in ms).
     */
    var totalTime: Long = 0L
        set(value) {
            field = abs(value)
        }

    fun addTotalTime(add: Long) {
        this.totalTime += abs(add)
    }

    /**
     * Total track time with speed (in ms).
     */
    var totalTimeMove: Long = 0L
        set(value) {
            field = abs(value)
        }

    fun addTotalTimeMove(add: Long) {
        this.totalTimeMove += abs(add)
    }

    /**
     * Get duration of the track
     */
    fun getTime(onlyWithMove: Boolean): Long {
        return if (onlyWithMove) {
            totalTimeMove
        } else {
            totalTime
        }
    }

    // SPEED

    /**
     * Maximal speed of this route (in m/s).
     */
    var speedMax: Float = 0.0f

    /**
     * Get average speed value for a track.
     *
     * @param onlyWithMove `true` to get AVG. speed only during movement
     * @return average speed in m/s
     */
    fun getSpeedAverage(onlyWithMove: Boolean): Float {
        val time = getTime(onlyWithMove) / 1000.0f
        return if (time > 0) {
            getLength(onlyWithMove) / time
        } else {
            0.0f
        }
    }

    // ELEVATION & CHANGES

    /**
     * Maximum altitude on the track (in m).
     */
    var altitudeMax: Float = 0.0f

    /**
     * Minimum altitude on the track (in m).
     */
    var altitudeMin: Float = 0.0f

    /**
     * Neutral grade (distance) (in m).
     */
    var eleNeutralDistance: Float = 0.0f

    fun addEleNeutralDistance(add: Float) {
        this.eleNeutralDistance += add
    }

    /**
     * Neutral grade (elevation) (in m).
     */
    var eleNeutralHeight: Float = 0.0f

    /**
     * Add elevation value for "neutral height"
     */
    fun addEleNeutralHeight(add: Float) {
        this.eleNeutralHeight += add
    }

    /**
     * Positive grade (distance)
     */
    var elePositiveDistance: Float = 0.0f

    fun addElePositiveDistance(add: Float) {
        this.elePositiveDistance += add
    }

    /**
     * Positive grade (elevation)
     */
    var elePositiveHeight: Float = 0.0f

    /**
     * Add elevation value for "positive height"
     */
    fun addElePositiveHeight(add: Float) {
        this.elePositiveHeight += add
    }

    /**
     * Negative grade (distance)
     */
    var eleNegativeDistance: Float = 0.0f

    fun addEleNegativeDistance(add: Float) {
        this.eleNegativeDistance += add
    }

    /**
     * Negative grade (elevation)
     */
    var eleNegativeHeight: Float = 0.0f

    /**
     * Add elevation value for "negative height"
     */
    fun addEleNegativeHeight(add: Float) {
        this.eleNegativeHeight += add
    }

    /**
     * Check if the track has some elevation values.
     */
    fun hasElevationValues(): Boolean {
        return (altitudeMin != Float.POSITIVE_INFINITY && altitudeMin != 0.0f)
                || (altitudeMax != Float.NEGATIVE_INFINITY && altitudeMax != 0.0f)
    }

    // CADENCE

    /**
     * Number of measured cadence revolutions.
     */
    private var cadenceNumber: Double = 0.0

    /**
     * Time during which were cadence measured.
     */
    private var cadenceTime: Long = 0

    /**
     * Get average value of cadence.
     *
     * @return average cadence value (
     */
    val cadenceAverage: Int
        get() = if (cadenceNumber > 0 && cadenceTime > 0L) {
            val minutes = cadenceTime / (60.0 * 1000.0)
            (cadenceNumber / minutes).toInt()
        } else {
            0
        }

    /**
     * Maximum cadence value.
     */
    var cadenceMax: Int = 0
        private set

    /**
     * Add measured cadence values.
     *
     * @param revMeasured   measured cadence value (in rpm)
     * @param revAvgSegment average revolutions value in segment (in rpm)
     * @param measureTime   time of segment (in millis)
     */
    fun addCadenceMeasure(revMeasured: Int, revAvgSegment: Int, measureTime: Long) {
        // store values
        val inMinutes = measureTime * 1.0 / (60 * 1000)
        val numOfRevolutions = revAvgSegment * inMinutes
        cadenceNumber += numOfRevolutions
        cadenceTime += measureTime

        // compute max. heart rate
        cadenceMax = max(cadenceMax, revMeasured)
    }

    // ENERGY

    /**
     * Burned energy (in joule).
     */
    var energy: Int = 0
        private set

    /**
     * Add burned energy.
     *
     * @param energy burned energy (in joule)
     */
    fun addEnergy(energy: Int) {
        this.energy += energy
    }

    // HEART RATE

    /**
     * Number of measured beats.
     */
    private var heartRateBeats: Double = 0.0

    /**
     * Time during which were beats measured.
     */
    private var heartRateTime: Long = 0L

    /**
     * Average value of heart rate.
     */
    val heartRateAverage: Int
        get() = if (heartRateBeats > 0 && heartRateTime > 0L) {
            val minutes = heartRateTime / (60.0 * 1000.0)
            (heartRateBeats / minutes).toInt()
        } else {
            0
        }

    /**
     * Maximum HRM value.
     */
    var heartRateMax: Int = 0
        private set

    /**
     * Add measured heart rate values.
     *
     * @param hrmMeasured   measured heart rate value (in bpm)
     * @param hrmAvgSegment average heart rate value in segment (in bpm)
     * @param measureTime   time of segment (in millis)
     */
    fun addHeartRateMeasure(hrmMeasured: Int, hrmAvgSegment: Int, measureTime: Long) {
        // store values
        val inMinutes = measureTime * 1.0 / (60 * 1000)
        val numOfBeats = hrmAvgSegment * inMinutes
        heartRateBeats += numOfBeats
        heartRateTime += measureTime

        // compute max. heart rate
        heartRateMax = max(heartRateMax, hrmMeasured)
    }

    // POWER

    /**
     * Sum of the average power counter (power measurement W * time in sec).
     */
    private var powerAvgSum: Double = 0.0

    /**
     * Time during which were power measured (in millis).
     */
    private var powerTime: Long = 0L

    /**
     * Average value of power (W).
     */
    val powerAverage: Int
        get() = if (powerAvgSum > 0.0 && powerTime > 0L) {
            val isSeconds = powerTime / 1000.0
            (powerAvgSum / isSeconds).toInt()
        } else {
            0
        }

    /**
     * Maximum power value (W).
     */
    var powerMax: Int = 0
        private set

    /**
     * Add measured power values.
     *
     * @param powerMeasured measured power value (in W)
     * @param powerAvgSegment average power value in segment (in W)
     * @param timeSegment time of segment (in millis)
     */
    fun addPowerMeasure(powerMeasured: Int, powerAvgSegment: Int, timeSegment: Long) {
        // store values
        val inSeconds = timeSegment / 1000.0
        powerAvgSum += powerAvgSegment * inSeconds
        powerTime += timeSegment
        powerMax = max(powerMax, powerMeasured)
    }

    // STRIDES

    /**
     * Total number of strides..
     */
    var numOfStrides: Int = 0

    // TEMPERATURE

    /**
     * Flag if temperature values are available.
     */
    var hasTemperature: Boolean = false

    /**
     * Minimum measured temperature (in °C).
     */
    var temperatureMin: Float = Float.POSITIVE_INFINITY

    /**
     * Maximum measured temperature (in °C).
     */
    var temperatureMax: Float = Float.NEGATIVE_INFINITY

    /**
     * Add temperature to current track statistics.
     */
    fun addTemperature(temp: Float) {
        hasTemperature = true
        temperatureMin = min(temperatureMin, temp)
        temperatureMax = max(temperatureMax, temp)
    }

    //*************************************************
    // OTHER TOOLS
    //*************************************************

    init {
        // other variables
        resetStatistics()
    }

    /**
     * Reset all parameters of the track statistics.
     */
    fun resetStatistics() {
        // basic statistics variables
        totalLength = 0.0f
        totalLengthMove = 0.0f
        totalTime = 0L
        totalTimeMove = 0L
        speedMax = 0.0f
        resetStatisticsAltitude()

        cadenceNumber = 0.0
        cadenceTime = 0L
        cadenceMax = 0
        energy = 0
        heartRateBeats = 0.0
        heartRateTime = 0L
        heartRateMax = 0
        numOfStrides = 0
        powerAvgSum = 0.0
        powerTime = 0L
        powerMax = 0
        hasTemperature = false
        temperatureMin = Float.POSITIVE_INFINITY
        temperatureMax = Float.NEGATIVE_INFINITY
    }

    /**
     * Reset only statistics related to elevation values.
     */
    fun resetStatisticsAltitude() {
        altitudeMax = Float.NEGATIVE_INFINITY
        altitudeMin = Float.POSITIVE_INFINITY

        // graph variables
        eleNeutralDistance = 0.0f
        eleNeutralHeight = 0.0f
        elePositiveDistance = 0.0f
        elePositiveHeight = 0.0f
        eleNegativeDistance = 0.0f
        eleNegativeHeight = 0.0f
    }

    /**
     * Function that allows to merge more statistics into one.
     *
     * @param stats second statistics that will be merged into this.
     */
    fun appendStatistics(stats: TrackStats) {
        this.numOfPoints += stats.numOfPoints
        this.startTime = min(startTime, stats.startTime)
        this.stopTime = max(stopTime, stats.stopTime)
        this.totalLength += stats.totalLength
        this.totalLengthMove += stats.totalLengthMove
        this.totalTime += stats.totalTime
        this.totalTimeMove += stats.totalTimeMove
        this.speedMax = max(speedMax, stats.speedMax)

        this.altitudeMax = max(altitudeMax, stats.altitudeMax)
        this.altitudeMin = min(altitudeMin, stats.altitudeMin)
        this.eleNeutralDistance += stats.eleNeutralDistance
        this.eleNeutralHeight += stats.eleNeutralHeight
        this.elePositiveDistance += stats.elePositiveDistance
        this.elePositiveHeight += stats.elePositiveHeight
        this.eleNegativeDistance += stats.eleNegativeDistance
        this.eleNegativeHeight += stats.eleNegativeHeight

        this.cadenceNumber += stats.cadenceNumber
        this.cadenceTime += stats.cadenceTime
        this.cadenceMax = max(this.cadenceMax, stats.cadenceMax)
        this.energy += stats.energy
        this.heartRateBeats += stats.heartRateBeats
        this.heartRateTime += stats.heartRateTime
        this.heartRateMax = max(this.heartRateMax, stats.heartRateMax)
        this.powerAvgSum += stats.powerAvgSum
        this.powerTime += stats.powerTime
        this.powerMax = max(this.powerMax, stats.powerMax)
        this.hasTemperature = this.hasTemperature || stats.hasTemperature
        this.temperatureMin = min(this.temperatureMin, stats.temperatureMin)
        this.temperatureMax = max(this.temperatureMax, stats.temperatureMax)
        this.numOfStrides += stats.numOfStrides
    }

    //*************************************************
    // STORABLE
    //*************************************************

    override fun getVersion(): Int {
        return 3
    }

    @Throws(IOException::class)
    override fun readObject(version: Int, dr: DataReaderBigEndian) {
        numOfPoints = dr.readInt()
        startTime = dr.readLong()
        stopTime = dr.readLong()

        // basic variables
        totalLength = dr.readFloat()
        totalLengthMove = dr.readFloat()
        totalTime = dr.readLong()
        totalTimeMove = dr.readLong()
        speedMax = dr.readFloat()

        // altitude values
        altitudeMax = dr.readFloat()
        altitudeMin = dr.readFloat()

        // elevation variables
        eleNeutralDistance = dr.readFloat()
        eleNeutralHeight = dr.readFloat()
        elePositiveDistance = dr.readFloat()
        elePositiveHeight = dr.readFloat()
        eleNegativeDistance = dr.readFloat()
        eleNegativeHeight = dr.readFloat()
        dr.readFloat() // eleTotalAbsDistance
        dr.readFloat() // eleTotalAbsHeight

        // V1
        if (version >= 1) {
            heartRateBeats = dr.readInt().toDouble()
            heartRateTime = dr.readLong()
            heartRateMax = dr.readInt()
            energy = dr.readInt()
        }

        // V2
        if (version >= 2) {
            heartRateBeats = dr.readDouble()
            cadenceNumber = dr.readDouble()
            cadenceTime = dr.readLong()
            cadenceMax = dr.readInt()
        }

        // V3
        if (version >= 3) {
            numOfStrides = dr.readInt()
        }

        // V4
        if (version >= 4) {
            powerAvgSum = dr.readDouble()
            powerTime = dr.readLong()
            powerMax = dr.readInt()
            hasTemperature = dr.readBoolean()
            temperatureMin = dr.readFloat()
            temperatureMax = dr.readFloat()
        }
    }

    @Throws(IOException::class)
    override fun writeObject(dw: DataWriterBigEndian) {
        dw.writeInt(numOfPoints)
        dw.writeLong(startTime)
        dw.writeLong(stopTime)

        // basic variables
        dw.writeFloat(totalLength)
        dw.writeFloat(totalLengthMove)
        dw.writeLong(totalTime)
        dw.writeLong(totalTimeMove)
        dw.writeFloat(speedMax)

        // altitude values
        dw.writeFloat(altitudeMax)
        dw.writeFloat(altitudeMin)

        // elevation variables
        dw.writeFloat(eleNeutralDistance)
        dw.writeFloat(eleNeutralHeight)
        dw.writeFloat(elePositiveDistance)
        dw.writeFloat(elePositiveHeight)
        dw.writeFloat(eleNegativeDistance)
        dw.writeFloat(eleNegativeHeight)
        dw.writeFloat(0.0f) // eleTotalAbsDistance
        dw.writeFloat(0.0f) // eleTotalAbsHeight

        // V1
        dw.writeInt(heartRateBeats.toInt())
        dw.writeLong(heartRateTime)
        dw.writeInt(heartRateMax)
        dw.writeInt(energy)

        // V2
        dw.writeDouble(heartRateBeats)
        dw.writeDouble(cadenceNumber)
        dw.writeLong(cadenceTime)
        dw.writeInt(cadenceMax)

        // V3
        dw.writeInt(numOfStrides)

        // V4
        dw.writeDouble(powerAvgSum)
        dw.writeLong(powerTime)
        dw.writeInt(powerMax)
        dw.writeBoolean(hasTemperature)
        dw.writeFloat(temperatureMin)
        dw.writeFloat(temperatureMax)
    }
}